src/encode.js

```diff
@@ -0,0 +1,71 @@
+'use strict'
+
+const { isObject } = require('./objects')
+const { isNonEmptyString } = require('./strings')
+
+const reusedSearchParams = new URLSearchParams()
+const reusedSearchParamKey = '_'
+const reusedSearchParamOffset = 2 // '_='.length
+
+const { encodeURIComponent } = globalThis
+
+function encodeNamespace(namespace) {
+    return isNonEmptyString(namespace)
+        ? encodeURIComponent(namespace)
+              .replace(/%3A/g, ':')
+              .replace(/%2F/g, '/')
+        : ''
+}
+
+function encodeQualifierParam(param) {
+    if (isNonEmptyString(param)) {
+        // Param key and value are encoded with `percentEncodeSet` of
+        // 'application/x-www-form-urlencoded' and `spaceAsPlus` of `true`.
+        // https://url.spec.whatwg.org/#urlencoded-serializing
+        reusedSearchParams.set(reusedSearchParamKey, param)
+        return replacePlusSignWithPercentEncodedSpace(
+            reusedSearchParams.toString().slice(reusedSearchParamOffset)
+        )
+    }
+    return ''
+}
+
+function encodeQualifiers(qualifiers) {
+    if (isObject(qualifiers)) {
+        // Sort this list of qualifier strings lexicographically.
+        const qualifiersKeys = Object.keys(qualifiers).sort()
+        const searchParams = new URLSearchParams()
+        for (let i = 0, { length } = qualifiersKeys; i < length; i += 1) {
+            const key = qualifiersKeys[i]
+            searchParams.set(key, qualifiers[key])
+        }
+        return replacePlusSignWithPercentEncodedSpace(searchParams.toString())
+    }
+    return ''
+}
+
+function encodeSubpath(subpath) {
+    return isNonEmptyString(subpath)
+        ? encodeURIComponent(subpath).replace(/%2F/g, '/')
+        : ''
+}
+
+function encodeVersion(version) {
+    return isNonEmptyString(version)
+        ? encodeURIComponent(version).replace(/%3A/g, ':').replace(/%2B/g, '+')
```

Comment:
This doesnt seem to be correct, Im not sure why some encodings are being reversed but from any source that I can read (between the spec and discussions/PRs in the spec repo e.g. https://github.com/package-url/purl-spec/issues/58#issuecomment-3009152400) versions are meant to be fully encoded @jdalton